debug-tools/binlog-event-blackhole/fetcher.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go-mysql/client"
+	"github.com/siddontang/go/sync2"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/pkg/log"
+)
+
+// registerSlave register a slave connection on the master.
+func registerSlave(addr, username, password string, serverID uint32) (*client.Conn, error) {
+	conn, err := client.Connect(addr, username, password, "", func(c *client.Conn) {
+	})
+	if err != nil {
+		return nil, errors.Annotate(err, "connect to the master")
+	}
+
+	// takes as an indication that the client is checksum-aware.
+	// ref https://dev.mysql.com/doc/refman/8.0/en/c-api-binary-log-functions.html.
+	_, err = conn.Execute(`SET @master_binlog_checksum='NONE'`)
+	if err != nil {
+		return nil, errors.Annotate(err, `SET @master_binlog_checksum='NONE'`)
+	}
+
+	conn.ResetSequence()
+	packet := registerSlaveCommand(username, password, serverID)
+	err = conn.WritePacket(packet)
+	if err != nil {
+		return nil, errors.Annotatef(err, "write COM_REGISTER_SLAVE packet %v", packet)
+	}
+
+	_, err = conn.ReadOKPacket()
+	if err != nil {
+		return nil, errors.Annotate(err, "read OK packet")
+	}
+
+	return conn, nil
+}
+
+// startSync starts to sync binlog event from <name, pos>.
+func startSync(conn *client.Conn, serverID uint32, name string, pos uint32) error {
+	conn.ResetSequence()
+	packet := dumpCommand(serverID, name, pos)
+	err := conn.WritePacket(packet)
+	return errors.Annotatef(err, "write COM_BINLOG_DUMP %v", packet)
+}
+
+// closeConn closes the connection to the master server.
+func closeConn(conn *client.Conn) error {
+	deadline := time.Now().Add(time.Millisecond)
+	err := conn.SetReadDeadline(deadline)
+	if err != nil {
+		return errors.Annotatef(err, "set connection read deadline to %v", deadline)
+	}
+
+	return errors.Trace(conn.Close())
+}
+
+// readEventsWithGoMySQL reads binlog events from the master server with `go-mysql` pkg.
+func readEventsWithGoMySQL(ctx context.Context, conn *client.Conn) (uint64, uint64, time.Duration, error) {
```

Comment:
the differences with `readEventsWithoutGoMySQL` is  the `"github.com/siddontang/go-mysql/client".Conn` use a buffer, I originally thought that it also contains more logic, such as decoding/parsing.  

Let's do this now, you can add the complete `go-mysql` logic later.

Reply:
Yeh. after doing some benchmark later, I think more (fine-grained) things we can do.